src/lib/prov/tpm2/tpm2_crypto_backend/tpm2_crypto_backend.cpp

```diff
@@ -0,0 +1,859 @@
+
+      // PK_Encryptor_EME does not provide a way to pass in an output buffer.
+      // TODO: provide an `.encrypt()` overload that accepts an output buffer.
+      Botan::PK_Encryptor_EME encryptor(pubkey, rng, "Raw");
+      const auto encrypted = encryptor.encrypt({out_buffer, padded_bytes}, rng);
+
+      // We abused the `out_buffer` to hold the result of the padding. Hence, we
+      // now have to copy the encrypted data over the padded plaintext data.
+      *out_size = encrypted.size();
+      Botan::copy_mem(std::span{out_buffer, *out_size}, encrypted);
+
+      return TSS2_RC_SUCCESS;
+   });
+   #else
+   BOTAN_UNUSED(pub_tpm_key, in_size, in_buffer, max_out_size, out_buffer, out_size, label, userdata);
+   return TSS2_ESYS_RC_NOT_IMPLEMENTED;
+   #endif
+}
+
+/** Computation of an ephemeral ECC key and shared secret Z.
+ *
+ * According to the description in TPM spec part 1 C 6.1 a shared secret
+ * between application and TPM is computed (ECDH). An ephemeral ECC key and a
+ * TPM key are used for the ECDH key exchange.
+ * @param[in] key The key to be used for ECDH key exchange.
+ * @param[in] max_out_size the max size for the output of the public key of the
+ *            computed ephemeral key.
+ * @param[out] Z The computed shared secret.
+ * @param[out] Q The public part of the ephemeral key in TPM format.
+ * @param[out] out_buffer The public part of the ephemeral key will be marshaled
+ *             to this buffer.
+ * @param[out] out_size The size of the marshaled output.
+ * @param[in,out] userdata information.
+ * @retval TSS2_RC_SUCCESS on success
+ * @retval USER_DEFINED user defined errors on failure.
+ */
+static TSS2_RC get_ecdh_point(TPM2B_PUBLIC* key,
+                              size_t max_out_size,
+                              TPM2B_ECC_PARAMETER* Z,
+                              TPMS_ECC_POINT* Q,
+                              BYTE* out_buffer,
+                              size_t* out_size,
+                              void* userdata) {
+   #if defined(BOTAN_HAS_TPM2_ECC_ADAPTER)
+   return thunk([&] {
+      BOTAN_UNUSED(userdata);
+      BOTAN_ASSERT_NONNULL(key);
+      BOTAN_ASSERT_NOMSG(key->publicArea.type == TPM2_ALG_ECC);
+
+      const auto curve_name = Botan::TPM2::curve_id_tss2_to_botan(key->publicArea.parameters.eccDetail.curveID);
+      const auto curve_order_byte_size =
+         Botan::TPM2::curve_id_order_byte_size(key->publicArea.parameters.eccDetail.curveID);
+      if(!curve_name || !curve_order_byte_size) {
+         return TSS2_ESYS_RC_NOT_IMPLEMENTED;  // ECC curve not supported
+      }
+
+      const auto domain = Botan::EC_Group::from_name(curve_name.value());
+
+      BOTAN_ASSERT_NOMSG(curve_order_byte_size.value() == domain.get_p_bytes());
+
+      Botan::AutoSeeded_RNG rng;  //TODO: What RNG to use?
+
+      // 1: Generate ephemeral key
+      const Botan::ECDH_PrivateKey eph_key(rng, domain);
```

Comment:
This would be a good place to use `Public_Key::generate_another()` if it is possible to load the public key from the provided data first.

Reply:
True, but I need to access `ECDH_PrivateKey::public_point()` so an additional cast would be necessary

Reply:
[Not necessarily](https://github.com/Rohde-Schwarz/botan/pull/14#discussion_r1789677706). But admittedly: we don't need the flexibility offered by `generate_another()` for the use case at hand.